### PR TITLE
Make topk/bottomk aggregators.

### PR DESCRIPTION
**Pull request by brian-brazil**
There'll be at least one more aggregator with a parameter, so I've added `isAggregatorWithParam` now rather than hardcoding in the places it's used and having to change later.

### CODE REVIEW

**Comment by fabxc** [comment]
We should validate the type of the first expression to be scalar. There is second phase in the parser where the AST is type-validated.

**Comment by brian-brazil** [comment]
The param isn't always going to be a scalar, for #622 it'll be a string.

**Comment by fabxc** [comment]
Still can be statically checked. For each function we define a clear type signature. With aggregators changing into the same direction, we should do the same thing here.

Runtime failures on something we can verify at parse time is a step backwards.

**Comment by brian-brazil** [comment]
That's added.

File: promql/engine.go [hunk 1]
```diff
@@ -1098,44 +1108,65 @@ func (ev *evaluator) aggregation(op itemType, grouping model.LabelNames, without
...
 			}
 			result[groupingKey] = &groupedAggregation{
 				labels:           m,
-				value:            sample.Value,
-				valuesSquaredSum: sample.Value * sample.Value,
+				value:            s.Value,
+				valuesSquaredSum: s.Value * s.Value,
 				groupCount:       1,
 			}
+			if op == itemTopK {
```
Comment (fabxc) [inline promql/engine.go]: This method/type is blowing up a fair bit now. Time for a refactoring.
If you are not keen to do it in here, could you file an issue?

**Comment by fabxc** [comment]
Functionality looks fine. Not too happy with the blow up of the aggregation code though.

File: promql/engine.go (hunk 1 above)
Comment (brian-brazil) [inline promql/engine.go]: It's getting messy alright, not sure how to clean it up. #1789 filed.

**Comment by fabxc** [comment]
👍

**Pull request merged**